### PR TITLE
[CUDA] dense_tensorcore/batch_matmul_tensorcore support int8/int4

### PR DESCRIPTION
**Pull request by wyc-ruiker**
Let dense_tensorcore and batch_matmul_tensorcore support int8/int4.
Before this pr, the [vision transform (vit)](https://arxiv.org/abs/2010.11929) latency (https://github.com/apache/tvm/pull/7814) in Tesla T4 is: 
vit int4: 4.71 ms
vit int8: 3.48 ms
After this pr:
vit int4: 2.93 ms
vit int8: 2.97 ms

@jcf94 @jwfromm @huochaitiantang  could you help review this pr?

### CODE REVIEW

**Comment by jcf94** [comment]
Thanks for your continue contribution on the tensor core schedule! @wyc-ruiker I'll help reivew when I have time.

p.s. Recently I added a new op `nn.matmul` which extend the `nn.dense` to support data tensor and weight tensor to be in transposed or non-transposed format. For a model from frameworks like TensorFlow, TVM will insert an extra transpose for `nn.dense` while use `nn.matmul` can get rid of this.
I'm not sure but maybe it will be beneficial to use it if your model suffered from some performance issue on the inserted transpose.

**Comment by wyc-ruiker** [comment]
> Thanks for your continue contribution on the tensor core schedule! @wyc-ruiker I'll help reivew when I have time.
> 
> p.s. Recently I added a new op `nn.matmul` which extend the `nn.dense` to support data tensor and weight tensor to be in transposed or non-transposed format. For a model from frameworks like TensorFlow, TVM will insert an extra transpose for `nn.dense` while use `nn.matmul` can get rid of that.
> I'm not sure but maybe it will be beneficial to use it if your model suffered from some performance issue on the inserted transpose.

```
  %1552 = reshape(%1551, newshape=[-1, 64, 50]) /* ty=Tensor[(12, 64, 50), float32] */;
  %1553 = transpose(%1552, axes=[0, 2, 1]) /* ty=Tensor[(12, 50, 64), float32] */;
  %1554 = multiply(%1553, 16f /* ty=float32 */) /* ty=Tensor[(12, 50, 64), float32] */;
  %1555 = round(%1554) /* ty=Tensor[(12, 50, 64), float32] */;
  %1556 = clip(%1555, a_min=-127f, a_max=127f) /* ty=Tensor[(12, 50, 64), float32] */;
  %1557 = cast(%1549, dtype="int8") /* ty=Tensor[(12, 50, 64), int8] */;
  %1558 = cast(%1556, dtype="int8") /* ty=Tensor[(12, 50, 64), int8] */;
  %1559 = nn.batch_matmul(%1557, %1558, meta[relay.attrs.BatchMatmulAttrs][61]) /* ty=Tensor[(12, 50, 50), int32] */;
```
Thanks, But in our vit network, it looks like we have some performance issues before `nn.batch_matmul`. Waiting for your adding full transpose support for `nn.batch_matmul`!

File: python/tvm/topi/cuda/batch_matmul_tensorcore.py [hunk 1]
```diff
@@ -94,32 +92,35 @@ def _schedule(cfg, s, C):
...
+                cfg.define_knob("wmma_m", [8, 16, 32])
+            wmma_k = 16
+            wmma_m = cfg["wmma_m"].val
+            if wmma_m == 16:
+                wmma_n = 16
+            elif wmma_m == 8:
+                wmma_n = 32
+            elif wmma_m == 32:
+                wmma_n = 8
+        else:
```
Comment (jcf94) [inline python/tvm/topi/cuda/batch_matmul_tensorcore.py:110]: Does the else branch for int4?
Even through we can assume the op strategy has already done the type check, still suggest better to specify it clearly in the code, then add an extra else branch to raise warnning.

File: python/tvm/topi/cuda/dense_tensorcore.py [hunk 2]
```diff
@@ -127,33 +131,36 @@ def _schedule_dense_tensorcore(cfg, s, C):
...
+            cfg.define_knob("wmma_m", [8, 16, 32])
+        wmma_k = 16
+        wmma_m = cfg["wmma_m"].val
+        if wmma_m == 16:
+            wmma_n = 16
+        elif wmma_m == 8:
+            wmma_n = 32
+        elif wmma_m == 32:
+            wmma_n = 8
+    else:
```
Comment (jcf94) [inline python/tvm/topi/cuda/dense_tensorcore.py:150]: ditto.

File: python/tvm/topi/cuda/tensorcore_alter_op.py [hunk 3]
```diff
@@ -70,31 +70,38 @@ def _batch_matmul_legalize(attrs, inputs, arg_types):
...
+
+    logger.info("batch_matmul pad_to_tensorcore, extra_flops %s", extra_flops)
+    if dm or dk:
+        x_ = relay.nn.pad(x, pad_width=((0, 0), (0, dm), (0, dk)))
+    else:
+        x_ = x
+    if dn or dk:
+        y_ = relay.nn.pad(y, pad_width=((0, 0), (0, dn), (0, dk)))
+    else:
+        y_ = y
```
Comment (jcf94) [inline python/tvm/topi/cuda/tensorcore_alter_op.py:97]:
```suggestion
    x_ = relay.nn.pad(x, pad_width=((0, 0), (0, dm), (0, dk))) if dm or dk else x
    y_ = relay.nn.pad(y, pad_width=((0, 0), (0, dn), (0, dk))) if dn or dk else y
```

File: python/tvm/topi/cuda/tensorcore_alter_op.py [hunk 4]
```diff
@@ -70,31 +70,38 @@ def _batch_matmul_legalize(attrs, inputs, arg_types):
...
+    out_ = relay.nn.batch_matmul(x_, y_, attrs.out_dtype)
+    if dm or dn:
+        original_out_shape = [x.value for x in output_tensor.shape]
+        out = relay.strided_slice(out_, begin=[0, 0, 0], end=original_out_shape)
+    else:
+        out = out_
+    return out
```
Comment (jcf94) [inline python/tvm/topi/cuda/tensorcore_alter_op.py:104]:
```suggestion
    out = relay.strided_slice(out_, begin=[0, 0, 0], end=[x.value for x in output_tensor.shape]) if dm or dn else out_
    return out
```

File: python/tvm/topi/cuda/tensorcore_alter_op.py [hunk 5]
```diff
@@ -147,30 +155,37 @@ def _dense_legalize(attrs, inputs, arg_types):
...
+        y_ = relay.nn.pad(y, pad_width=((0, dn), (0, dk)))
+    else:
+        y_ = y
+    out_ = relay.nn.dense(x_, y_, **new_attrs)
+    if dm or dn:
+        original_out_shape = [x.value for x in output_tensor.shape]
+        out = relay.strided_slice(out_, begin=[0, 0], end=original_out_shape)
+    else:
+        out = out_
+    return out
```
Comment (jcf94) [inline python/tvm/topi/cuda/tensorcore_alter_op.py:188]: ditto. Refine these lines.

**Review by jcf94 — COMMENTED** [review]
Thanks! @wyc-ruiker Overall looks great to me.

Just some nit-pick.

File: python/tvm/topi/cuda/batch_matmul_tensorcore.py (hunk 1 above)
Comment (wyc-ruiker) [reply python/tvm/topi/cuda/batch_matmul_tensorcore.py]: done.

File: python/tvm/topi/cuda/dense_tensorcore.py (hunk 2 above)
Comment (wyc-ruiker) [reply python/tvm/topi/cuda/dense_tensorcore.py]: done.

File: python/tvm/topi/cuda/tensorcore_alter_op.py (hunk 5 above)
Comment (wyc-ruiker) [reply python/tvm/topi/cuda/tensorcore_alter_op.py]: done.

**Review by jcf94 — APPROVED** [review]
Thanks! @wyc-ruiker

**Comment by jcf94** [comment]
Push agian to re-triggle the CI? @wyc-ruiker

**Pull request merged**